addons/controllers/cpi/cpiconfig_controller.go

```diff
@@ -0,0 +1,314 @@
+// Copyright 2022 VMware, Inc. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+// Package controllers implements k8s controller functionality for vsphere-cpi.
+package controllers
+
+import (
+	"context"
+	"fmt"
+
+	clusterapiutil "sigs.k8s.io/cluster-api/util"
+
+	capvv1beta1 "sigs.k8s.io/cluster-api-provider-vsphere/api/v1beta1"
+
+	"github.com/pkg/errors"
+
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+
+	"github.com/go-logr/logr"
+
+	"github.com/vmware-tanzu/tanzu-framework/addons/pkg/constants"
+	"github.com/vmware-tanzu/tanzu-framework/addons/pkg/util"
+	cpiv1alpha1 "github.com/vmware-tanzu/tanzu-framework/apis/cpi/v1alpha1"
+
+	yaml "gopkg.in/yaml.v3"
+	v1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+
+	clusterapiv1beta1 "sigs.k8s.io/cluster-api/api/v1beta1"
+	clusterapipatchutil "sigs.k8s.io/cluster-api/util/patch"
+
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+)
+
+const (
+	CPINonParavirtualMode = "vsphereCPI"
+	CPIParavirtualMode    = "vsphereParavirtualCPI"
+)
+
+// CPIConfigReconciler reconciles a CPIConfig object
+type CPIConfigReconciler struct {
+	client.Client
+	Log    logr.Logger
+	Scheme *runtime.Scheme
+}
+
+//+kubebuilder:rbac:groups=cpi.tanzu.vmware.com,resources=cpiconfigs,verbs=get;list;watch;create;update;patch;delete
+//+kubebuilder:rbac:groups=cpi.tanzu.vmware.com,resources=cpiconfigs/status,verbs=get;update;patch
+
+// Reconcile the CPIConfig CRD
+func (r *CPIConfigReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	r.Log = r.Log.WithValues("CPIConfig", req.NamespacedName)
+
+	r.Log.Info("Start reconciliation for CPIConfig")
+
+	// fetch CPIConfig resource
+	cpiConfig := &cpiv1alpha1.CPIConfig{}
+	if err := r.Client.Get(ctx, req.NamespacedName, cpiConfig); err != nil {
+		if apierrors.IsNotFound(err) {
+			r.Log.Info("CPIConfig resource not found")
+			r.Log.Info("CPIConfig resource not found")
+			return ctrl.Result{}, nil
+		}
+		r.Log.Error(err, "Unable to fetch CPIConfig resource")
+		return ctrl.Result{}, err
+	}
+
+	cpiConfig = cpiConfig.DeepCopy()
+	cluster, err := r.getOwnerCluster(ctx, cpiConfig)
+	if apierrors.IsNotFound(err) {
+		return ctrl.Result{}, nil
+	} else if err != nil {
+		return ctrl.Result{}, err
+	}
+	if res, err := r.ReconcileCPIConfig(ctx, req, cpiConfig, cluster); err != nil {
```

Comment:
Why do you need to pass ```req``` to this function?

Reply:
removed